src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -350,16 +351,20 @@ private DocumentState SetParseOptions(ParseOptions options, bool onlyPreprocesso
 
             ValueSource<TreeAndVersion>? newTreeSource = null;
 
-            // Optimization: if we are only changing preprocessor directives, and we've already parsed the existing tree and it didn't have
-            // any, we can avoid a reparse since the tree will be parsed the same.
+            // Optimization: if we are only changing preprocessor directives, and we've already parsed the existing tree
+            // and it didn't have any, we can avoid a reparse since the tree will be parsed the same.
+            //
+            // We only need to care about `#if` directives as those are the only sorts of directives that can affect how
+            // a tree is parsed.
             if (onlyPreprocessorDirectiveChange &&
                 _treeSource.TryGetValue(out var existingTreeAndVersion))
             {
                 var existingTree = existingTreeAndVersion.Tree;
 
                 SyntaxTree? newTree = null;
 
-                if (existingTree.TryGetRoot(out var existingRoot) && !existingRoot.ContainsDirectives)
+                var syntaxKinds = _languageServices.GetRequiredService<ISyntaxKindsService>();
+                if (existingTree.TryGetRoot(out var existingRoot) && !existingRoot.ContainsDirective(syntaxKinds.IfDirectiveTrivia))
```

Comment:
utilizes new compiler API we added for this exact purpose.  `.ContainsDirectives` is too broad, firing even for things that do not affect parsing.  `.ContainsDirective(SyntaxKind.IfDirectiveTrivia)` means we only fire on trees hat could genuinely parse differently.

Reply:
Is there any broken code scenarios we need to worry about, like there's not an #if but there's an #elif that still results in different trees? I sure don't care about optimizing this case; this is entirely a "I hope we don't crash" question.